ffi/derive/src/impl_visitor.rs

```diff
@@ -93,7 +104,9 @@ pub struct FnDescriptor<'ast> {
     pub output_arg: Option<Arg>,
 }
 
-struct ImplVisitor<'ast> {
+struct ImplVisitor<'ast, 'emitter> {
+    emitter: &'emitter mut Emitter,
+    fatal: bool,
```

Comment:
I am not sure how a type-level guard for this might look like.. Can you give an example?

Reply:
I mean, to express the invariant on a type-level. For example, instead of having a `fatal: bool` flag, `ImplVisitor` might be a enum:

```rust
enum ImplVisitor<'ast, 'emitter> {
    Fatal,
    Valid {
        emitter: &'emitter mut Emitter,
        attrs: Vec<&'ast Attribute>,
        trait_name: Option<&'ast Path>,
        /// Resolved type of the `Self` type
        self_ty: Option<&'ast Path>,
        associated_types: Vec<(&'ast Ident, &'ast Type)>,
        fns: Vec<FnDescriptor<'ast>>,
    }
}
```

So that when `ImplDescriptor` is created from `ImplVisitor`, it is "impossible" to dismiss its `Fatal` state.

Reply:
It would complicate all the other visitor methods though, because they would have to pattern match to get access to the fields... Not sure it's worth it

Reply:
> Not sure it's worth it

Me as well.